pkg/informer/informer.go

```diff
@@ -645,3 +679,142 @@ func (d *DynamicDiscoverySharedInformerFactory) Unsubscribe(id string) {
 
 	delete(d.subscribers, id)
 }
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerGroups() (*metav1.APIGroupList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	groups := make([]metav1.APIGroup, len(d.discoveryData))
+	failedGroups := make(map[schema.GroupVersion]error)
+
+	for i, apiResourceList := range d.discoveryData {
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			failedGroups[schema.GroupVersion{Group: apiResourceList.GroupVersion}] = err
+			continue
+		}
+		groups[i] = metav1.APIGroup{
+			Name: gv.Group,
+			Versions: []metav1.GroupVersionForDiscovery{
+				{
+					GroupVersion: gv.String(),
+					Version:      gv.Version,
+				},
+			},
+			PreferredVersion: metav1.GroupVersionForDiscovery{
+				GroupVersion: gv.String(),
+				Version:      gv.Version,
+			},
+		}
+	}
+
+	if len(failedGroups) > 0 {
+		return &metav1.APIGroupList{Groups: groups}, &discovery.ErrGroupDiscoveryFailed{Groups: failedGroups}
+	}
+
+	return &metav1.APIGroupList{Groups: groups}, nil
+}
+
+var _ discovery.DiscoveryInterface = &DynamicDiscoverySharedInformerFactory{}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	for _, apiResourceList := range d.discoveryData {
+		if apiResourceList.GroupVersion == groupVersion {
+			return apiResourceList.DeepCopy(), nil
+		}
+	}
+
+	// ignore 403 or 404 error to be compatible with a v1.0 server.
+	if groupVersion == "v1" {
+		return &metav1.APIResourceList{GroupVersion: groupVersion}, nil
+	}
+
+	return nil, errors.NewNotFound(schema.GroupResource{Group: groupVersion}, "")
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+	defer d.informersLock.RUnlock()
+
+	retGroups := make([]*metav1.APIGroup, len(d.discoveryData))
+	failedGroups := make(map[schema.GroupVersion]error)
+	retResourceList := make([]*metav1.APIResourceList, len(d.discoveryData))
+
+	for i, apiResourceList := range d.discoveryData {
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
```

Comment:
Why do we do this parsing & transformation when fulfilling a discovery request, as opposed to filling our in-memory cache (`d.discoveryData`) with the data we're going to need to serve and making the per-request work `O(1)`?

Reply:
Is this outstanding?

Reply:
I've improved this at the same time I've addressed the removal of `rest.Interface`, so discovery and REST mapping methods share the same data structures.